subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/ChildMapFactory.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+public class ChildMapFactory {
+    /**
+     * If a node has fewer children, we use a linear search for the child.
+     * We use this limit since {@link VfsRelativePath#compareToFirstSegment(String, CaseSensitivity)}
+     * is about twice as slow as {@link VfsRelativePath#hasPrefix(String, CaseSensitivity)},
+     * so comparing the searched path to all of the children is actually faster than doing a binary search.
+     */
+    private static final int MINIMUM_CHILD_COUNT_FOR_BINARY_SEARCH = 10;
+
+    public static <T> ChildMap<T> childMap(List<ChildMap.Entry<T>> sortedEntries) {
```

Comment:
I think there should be a variant of this taking `Collection`.

Reply:
I added a method. The method also sorts the children accordingly to the case sensitivity.